File: LeetCode/1046/Solution.cpp
class Solution {
public:
    int lastStoneWeight(vector<int>& stones) {
        priority_queue<int> pq;
        
        for(int stone : stones)
            pq.push(stone);
        
        while(pq.size() > 1) {
            int x = pq.top();
            pq.pop();
            
            int y = pq.top();
            pq.pop();
            
            if(x == y)
                continue;
            else if(x > y)
                pq.push(x - y);
            else if(x < y)
                pq.push(y - x);
        }
        
        if(pq.size() == 0)
            return 0;
        else
            return pq.top();
    }
};

File: LeetCode/1046/Solution.kt
class Solution {
    fun lastStoneWeight(stones: IntArray): Int {
        val pq = PriorityQueue<Int>(Collections.reverseOrder())
        stones.forEach {
            pq.offer(it)
        }
        while(pq.size >= 2) {
            val a = pq.poll()
            val b = pq.poll()

            if(a > b) {
                pq.offer(a-b)
            }
        }

        return pq.poll() ?: 0
    }
}
